### PR TITLE
Merge 0.5

### PR DESCRIPTION
**Pull request by pva701**
:nate:

### CODE REVIEW

File: core/Pos/Binary/Crypto.hs [hunk 1]
```diff
@@ -142,7 +142,7 @@ secretKeyLength, publicKeyLength, signatureLength, chainCodeLength,
     encryptedKeyLength, passphraseLength :: Int
 secretKeyLength = 32
 publicKeyLength = 32
-encryptedKeyLength = 128
+encryptedKeyLength = 96
```
Comment (gromakovsky) [inline core/Pos/Binary/Crypto.hs]: That's the exact the reason why this merge is bad.

File: core/Pos/Util/Modifier.hs [hunk 2]
```diff
@@ -53,6 +61,11 @@ instance (Eq k, Hashable k) =>
 
 instance (Eq k, Hashable k) => Semigroup (MapModifier k v)
 
+instance (Buildable k, Buildable v) => Buildable (MapModifier k v) where
+    build mm =
+      bprint ("MapModifier { deletions "%listJson%", insertions: "%listJson%"}")
+      (deletions mm) (map (bprint pairF) (insertions mm))
```
Comment (gromakovsky) [inline core/Pos/Util/Modifier.hs]: I don't know who was adding this instance, but FYI: there is `mapBuilder` and `mapJson` formatter.

File: lwallet/Pos/Wallet/Light/Redirect.hs [hunk 3]
```diff
@@ -40,8 +41,8 @@ getBalanceWallet = getBalanceFromUtxo
 
 getBlockHistoryWallet
     :: (MonadReader ctx m, HasLens LWS.WalletState ctx LWS.WalletState, MonadIO m)
-    => [Address] -> m (DList TxHistoryEntry)
-getBlockHistoryWallet addrs = do
+    => Tagged ssc ([Address] -> m (DList TxHistoryEntry))
+getBlockHistoryWallet = Tagged $ \addrs -> do
```
Comment (flyingleafe) [inline lwallet/Pos/Wallet/Light/Redirect.hs]: No, I was so happy when I've seen that we can get rid of this `Tagged` shit here in master, please remove it =\\

File: src/Pos/Wallet/Light/Redirect.hs [hunk 4]
```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE TypeFamilies #-}
+
+-- | Various monadic redirects implementing wallet constraints. Mostly stubs.
+
+module Pos.Wallet.Light.Redirect
```
Comment (gromakovsky) [inline src/Pos/Wallet/Light/Redirect.hs]: Wtf is that?
1. Why isn't it in `lwallet`?
2. We don't use redirects nowadays.

File: core/Pos/Binary/Crypto.hs (hunk 1 above)
Comment (pva701) [inline core/Pos/Binary/Crypto.hs]: I AM BAD BOY!

File: src/Pos/Client/Txp/History.hs [hunk 5]
```diff
@@ -36,8 +37,11 @@ import           Data.Coerce                  (coerce)
 import           Data.DList                   (DList)
 import qualified Data.DList                   as DL
 import qualified Data.Map.Strict              as M (lookup)
+import           Data.Tagged                  (Tagged (..))
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/History.hs]: Same shit: enable `AllowAmbiguousTypes` and get rid of `Tagged`

File: src/Pos/Client/Txp/History.hs [hunk 6]
```diff
@@ -176,18 +186,28 @@ instance (Monad m, MonadReader ctx m, HasLens GenesisUtxo ctx GenesisUtxo) =>
 -- | A class which have methods to get transaction history
 class (Monad m, SscHelpersClass ssc) => MonadTxHistory ssc m | m -> ssc where
     getBlockHistory
-        :: [Address] -> m (DList TxHistoryEntry)
+        :: SscHelpersClass ssc
+        => Tagged ssc ([Address] -> m (DList TxHistoryEntry))
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/History.hs]: Same here

File: src/Pos/Client/Txp/History.hs [hunk 7]
```diff
@@ -234,8 +254,11 @@ getLocalHistoryDefault
 getLocalHistoryDefault addrs = runDBToil . evalToilTEmpty $ do
     let mapper (txid, TxAux {..}) =
             (WithHash taTx txid, taWitness, taDistribution)
-    ltxs <- getLocalTxs
-    txs <- getRelatedTxsByAddrs addrs Nothing Nothing $ map mapper ltxs
+        topsortErr = TxpInternalError
+            "getLocalHistory: transactions couldn't be topsorted!"
+    ltxs <- map mapper <$> getLocalTxs
+    txs <- getRelatedTxsByAddrs addrs Nothing Nothing =<<
+            maybeThrow topsortErr (topsortTxs (view _1) ltxs)
```
Comment (flyingleafe) [inline src/Pos/Client/Txp/History.hs]: From where this extra whitespace came from?... 😿

**Comment by pva701** [comment]
@flyingleafe it's for you :gun: 
@gromakovsky and it's for you :hocho:

**Comment by pva701** [comment]
>src/Pos.Wallet.Light.Redirect should not exist.

the best form of existence is nonexistence, right?

File: src/Pos/Wallet/Light/Redirect.hs (hunk 4 above)
Comment (pva701) [inline src/Pos/Wallet/Light/Redirect.hs]: @gromakovsky I am from 2k16

**Pull request merged**